src/nbla/cuda/function/generic/batch_inv.cu

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <nbla/array.hpp>
+#include <nbla/cuda/array/cuda_array.hpp>
+#include <nbla/cuda/common.hpp>
+#include <nbla/cuda/function/batch_inv.hpp>
+#include <nbla/function/batch_matmul.hpp>
+#include <nbla/cuda/math.hpp>
+#include <nbla/variable.hpp>
+#include <iostream>
+
+namespace nbla {
+
+// Sets head pointers of matrices in mini-batch.
+template <typename T>
+__global__ void kernel_set_batch_pointers(int batchSize, int n, const T **ptr,
+                                          const T *head) {
+  NBLA_CUDA_KERNEL_LOOP(idx, batchSize) { ptr[idx] = head + idx * n * n; }
+}
+
+// A macro that creates an array of pointers of matrices.
+#define NBLA_GET_BATCH_POINTERS(PTR, NAME, BATCH, CONST)                       \
+  CudaCachedArray list_##PTR(sizeof(Tc *) * BATCH, dtypes::BYTE,               \
+                             this->ctx_);                                      \
+  CONST Tc **dev_list_##NAME =                                                 \
+      reinterpret_cast<CONST Tc **>(list_##PTR.pointer<void>());               \
+  NBLA_CUDA_LAUNCH_KERNEL_SIMPLE(kernel_set_batch_pointers, BATCH, dim_,       \
+                                 (const T **)dev_list_##NAME, (const T *)PTR)
+
+// ----------------------------------------------------------------------
+// With cublas<t>getrfBatched and cublas<t>getriBatched
+// ----------------------------------------------------------------------
+template <typename T>
+void BatchInvCuda<T>::setup_impl(const Variables &inputs,
+                                      const Variables &outputs) {
+  BatchInv<T>::setup_impl(inputs, outputs);
+  batch_size_ = inputs[0]->shape()[0];
+  dim_ = inputs[0]->shape()[1];
+}
+
+template <typename T>
+void BatchInvCuda<T>::forward_impl(const Variables &inputs,
+                                      const Variables &outputs) {
+  cuda_set_device(this->device_);
+  const Tc *x = inputs[0]->get_data_pointer<Tc>(this->ctx_);
+  Tc *y = outputs[0]->cast_data_and_get_pointer<Tc>(this->ctx_, true);
+
+  shared_ptr<CudaCachedArray> pivot =
```

Comment:
You don't have to use shared pointers (I know it is used in other functions though...)